pkg/resource/deploy/source_eval.go

```diff
@@ -1639,10 +1650,108 @@ func (rm *resmon) SignalAndWaitForShutdown(ctx context.Context, req *emptypb.Emp
 	return &emptypb.Empty{}, nil
 }
 
+func (rm *resmon) wrapResourceHookCallback(name string, cb *pulumirpc.Callback) (ResourceHookFunction, error) {
+	client, err := rm.GetCallbacksClient(cb.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(ctx context.Context, urn resource.URN, id resource.ID,
+		newInputs, oldInputs, newOutputs, oldOutputs resource.PropertyMap,
+	) error {
+		logging.V(6).Infof("ResourceHook calling hook %q for urn %s", name, urn)
+		var mNewInputs, mOldInputs, mNewOutputs, mOldOutputs *structpb.Struct
+		if newInputs != nil {
+			mNewInputs, err = plugin.MarshalProperties(newInputs, plugin.MarshalOptions{
```

Comment:
All these MarshalOptions are the same, just pull out to an `mopts` variable at the start.